File: mt-user/src/main/kotlin/com/mt/mtuser/common/page/PageView.java
package com.mt.mtuser.common.page;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

/**
 * Created by gyh on 2018/10/19.
 * 分页对象
 */
@Deprecated
public class PageView<T> {
    private Logger logger = LoggerFactory.getLogger(this.getClass().getSimpleName());
    // 当前页号
    private int pageNum;

    // 每页的数量
    private int pageSize;

    // 总记录数
    private long total;

    // 总页数
    private int pages;

    // 结果集
    private List<T> list;

    public PageView(PageQuery pageQuery) {
        this.pageNum = pageQuery.getPageNum();
        this.pageSize = pageQuery.getPageSize();
    }

    public PageView() {
    }

    public void count(PageQuery query) {
        String sql = query.queryBuild.build().trim();
        String select = "select count(1) " + sql.substring(sql.lastIndexOf("from"));
        logger.info(select);
    }

    public void count(String sql) {
        String select = "select count(1) " + sql.substring(sql.lastIndexOf("from"));
        logger.info(select);
    }

    public static int getPages(long total, int pageSize) {
        if (total == 0 || pageSize == 0) {
            return 0;
        }
        return (int) (total % pageSize == 0 ? (total / pageSize) : (total / pageSize + 1));
    }

    public int getPages() {
        return getPages(this.total, this.pageSize);
    }

    public int getPageNum() {
        return pageNum;
    }

    public void setPageNum(int pageNum) {
        this.pageNum = pageNum;
    }

    public int getPageSize() {
        return pageSize;
    }

    public void setPageSize(int pageSize) {
        this.pageSize = pageSize;
    }

    public long getTotal() {
        return total;
    }

    public void setTotal(long total) {
        this.total = total;
    }

    public void setPages(int pages) {
        this.pages = pages;
    }

    public List<T> getList() {
        return list;
    }

    public void setList(List<T> list) {
        this.list = list;
    }
}

File: mt-user/src/main/kotlin/com/mt/mtuser/entity/page/PageView.kt
package com.mt.mtuser.entity.page

import kotlinx.coroutines.reactive.awaitSingle
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.r2dbc.core.DatabaseClient
import org.springframework.data.relational.core.mapping.Table
import org.springframework.data.relational.core.query.Criteria
import reactor.core.publisher.Flux
import kotlin.reflect.full.findAnnotation

/**
 * Created by gyh on 2020/3/21.
 */
class PageView<T : Any> {
    // 当前页号
    var pageNum = 0

    // 每页的数量
    var pageSize = 0

    // 总记录数
    var total: Long? = 0
    var item: List<T>? = null
}

val logger: Logger = LoggerFactory.getLogger(PageView::class.java)

/**
 * 如果指定了[where] 使用[where], 放弃使用[pageQuery.where()],如果没有指定[where]，则使用[pageQuery.where()]作为条件
 */
suspend inline fun <reified T : Any> getPage(data: Flux<T>, connect: DatabaseClient, pageQuery: PageQuery, where: Criteria? = null): PageView<T> {
    val tableName = T::class.findAnnotation<Table>()?.value ?: T::class.simpleName
    ?: throw IllegalStateException("不支持的匿名类 ${T::class}")
    return getPage(data, connect, pageQuery, tableName, where)
}

suspend fun <T: Any> getPage(data: Flux<T>, connect: DatabaseClient, pageQuery: PageQuery, tableName: String, where: Criteria? = null): PageView<T> {
    val pageView = PageView<T>()
    var sqlWhere = (where?.toString() ?: pageQuery.where().toString())
    if (sqlWhere.isNotBlank()) {
        sqlWhere = " where $sqlWhere"
    }
    logger.info(sqlWhere)
    val count = connect.execute("select count(1) from $tableName $sqlWhere")
            .map { r, _ -> r.get(0, java.lang.Long::class.java) }.one().awaitSingle()
    pageView.total = count?.toLong()
    pageView.item = data.collectList().awaitSingle()
    pageView.pageNum = pageQuery.pageNum
    pageView.pageSize = pageQuery.pageSize
    return pageView
}

suspend fun  <T: Any> getPage(data: Flux<T>, connect: DatabaseClient, pageQuery: PageQuery, from: String, where: String): PageView<T> {
    val pageView = PageView<T>()
    val count = connect.execute("select count(1) from $from where $where")
            .map { r, _ -> r.get(0, java.lang.Long::class.java) }.one().awaitSingle()
    pageView.total = count?.toLong()
    pageView.item = data.collectList().awaitSingle()
    pageView.pageNum = pageQuery.pageNum
    pageView.pageSize = pageQuery.pageSize
    return pageView
}

